common/cache.go

```diff
@@ -10,87 +13,169 @@ import (
+	if err != nil {
+		Log.Fatal("open ", c.lengthsName, " failed: ", err)
+	}
+	lengths, err := ioutil.ReadFile(c.lengthsName)
+	if err != nil {
+		Log.Fatal("read ", c.lengthsName, " failed: ", err)
+	}
+	// the last entry in starts[] is where to write the next block
+	var offset int64
+	c.starts = append(c.starts, 0)
+	for i := 0; i < len(lengths)/4; i++ {
+		length := binary.LittleEndian.Uint32(lengths[i*4 : i*4+4])
+		offset += int64(length)
+		c.starts = append(c.starts, offset)
+		c.NextBlock++
+	}
+	if c.NextBlock > c.FirstBlock {
+		// There is at least one block; get the last block's hash
+		b := make([]byte, c.blockLength(c.NextBlock-1))
+		_, err := c.blocksFile.ReadAt(b, c.starts[c.NextBlock-c.FirstBlock-1])
+		if err != nil {
+			Log.Fatal("blocks read failed: ", err)
+		}
+		block := &walletrpc.CompactBlock{}
+		err = proto.Unmarshal(b, block)
+		if err != nil {
+			println("Error unmarshalling compact block")
+			return nil
+		}
+		c.LatestHash = make([]byte, len(block.Hash))
+		copy(c.LatestHash, block.Hash)
 	}
+	return c
+}
+
+func fileNames(chainName string) (string, string) {
+	return fmt.Sprintf("db-%s-lengths", chainName),
+		fmt.Sprintf("db-%s-blocks", chainName)
+}
+
+// only used for testing to ensure we're not using files from a previous run
+func CacheTestClean(chainName string) {
+	lengthsName, blockName := fileNames(chainName)
+	os.Remove(lengthsName)
+	os.Remove(blockName)
 }
 
 // Add adds the given block to the cache at the given height, returning true
 // if a reorg was detected.
-func (c *BlockCache) Add(height int, block *walletrpc.CompactBlock) (bool, error) {
-	// Invariant: m[firstBlock..nextBlock) are valid.
+func (c *BlockCache) Add(block *walletrpc.CompactBlock) error {
+	// Invariant: m[FirstBlock..NextBlock) are valid.
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
 
-	if height > c.nextBlock {
-		// restarting the cache (never happens currently), or first time
-		for i := c.firstBlock; i < c.nextBlock; i++ {
-			delete(c.m, i)
-		}
-		c.firstBlock = height
-		c.nextBlock = height
+	height := int(block.Height)
+	if height != c.NextBlock {
+		Log.Fatalf("bug, cache.Add non-consecutive blocks: height: %v, ccache: %+v", height, c)
 	}
-	// Invariant: m[firstBlock..nextBlock) are valid.
 
-	// If we already have this block, a reorg must have occurred;
-	// this block (and all higher) must be re-added.
-	h := height
-	if h < c.firstBlock {
-		h = c.firstBlock
+	// Add the new block and its length to the db files
+	data, err := proto.Marshal(block)
+	if err != nil {
+		return err
 	}
-	for i := h; i < c.nextBlock; i++ {
-		delete(c.m, i)
+	_, err = c.blocksFile.Write(data)
+	if err != nil {
+		Log.Fatal("blocks write failed: ", err)
 	}
-	c.nextBlock = height
-	if c.firstBlock > c.nextBlock {
-		c.firstBlock = c.nextBlock
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, uint32(len(data)))
+	_, err = c.lengthsFile.Write(b)
+	if err != nil {
+		Log.Fatal("lengths write failed: ", err)
```

Comment:
Bug: At this point a block has been written to the block file, but a length has not been added to the length file. `NewBlockCache` will not detect the problem, because it uses the number of entries in the length file to determine how many blocks to read from the blocks file. `NewBlockCache` doesn't check that there aren't *extra* blocks in the file.

The following could happen:

1. We try to add a block A, at height N, and reach this path due to some transient error. The blocks file has N blocks ending in A, but the length file only has N-1 lengths.
2. Lightwalletd gets restarted, `NewBlockCache` doesn't notice the error, lightwalletd serves blocks up to N-1.
3. We try to add a block B, at height N (say there was a reorg), but the transient error is resolved and this code works. Now, before lightwalletd restarts, it'll serve block B at height N, but after it restarts, it will serve block A at height N.

(Assuming blocks A and B are the same length, otherwise it might crash upon trying to read the block)

One way to fix this would be to add a "transaction in progress" flag to each file, and modify them by:

1. Set the "transaction in progress" flag in both files. Sync to guarantee that has been written to disk.
2. Write the block contents and then the length. Sync to guarantee it has all been written to disk.
3. Clear the "transaction in progress" flag in both files.
4. Upon loading the file, make sure both "transaction in progress" flags are cleared, or assume the database is corrupted.

(I guess it could be a flag in its own file or in just one of the files).

Reply:
> `NewBlockCache` doesn't check that there aren't _extra_ blocks in the file
What if `NewBlockCache` unconditionally just truncates the blocks file to the length it should have based on the lengths file? Would that fix this? 

> but after it restarts, it will serve block A at height N.
Is that because block B didn't get synced to disk? One thing I was trying to avoid is syncing files in `Add()` because while we're syncing with the blockchain, we're calling `Add()` for each block, so performance matters. (But notice we do `Sync()` before sleeping in `BlockIngestor()`.) I guess I'm thinking that with the checksum improvement, any corruption that occurs (that could have been prevented by syncing) will be detected and the bad data overwritten (by fetching from `zcashd` again).

Hmm, here's a thought: When we restart, instead of fetching from exactly where (we think) we left off, back up 100 blocks or something (go through exactly the reorg code), and resume from there. The time to re-download 100 blocks is very small. This would overwrite any corruption near the end of the files, which is by far the most likely place for it to be (because that's where writing occurs).